pkg/cloud/externalconn/impl_registry.go

```diff
@@ -138,11 +164,73 @@ func ExternalConnectionFromURI(
 		return nil, errors.Newf("no parseAndValidateFn found for external connection provider %s", externalConnectionURI.Scheme)
 	}
 
-	return parseAndValidateFn.parseAndValidateURI(ctx, execCfg, user, externalConnectionURI, defaultValidation)
+	return parseAndValidateFn.parseAndValidateURI(ctx, env, externalConnectionURI, defaultValidation)
 }
 
-// ValidationFn is the type taken by Register*Validation. Validation functions can ping the external connection.
-type ValidationFn func(ctx context.Context, execCfg interface{}, user username.SQLUsername, uri string) error
+type ExternalConnEnv struct {
+	// Settings refers to the cluster settings that apply to the BackupKMSEnv.
+	Settings *cluster.Settings
+	// Conf represents the ExternalIODirConfig that applies to the BackupKMSEnv.
+	Conf *base.ExternalIODirConfig
+	// DB is the database handle that applies to the BackupKMSEnv.
+	Isqldb isql.DB
+	// Username is the user that applies to the BackupKMSEnv.
+	Username username.SQLUsername
+	// ExternalConnectionTestingKnobs is used for testing.
+	ExternalConnectionTestingKnobs cloud.ExternalConnTestingKnobs
+	// ServerCfg is used to validate the external connection sink URI.
+	ServerCfg *execinfra.ServerConfig
+}
+
+func MakeExternalConnEnv(
+	settings *cluster.Settings,
+	conf *base.ExternalIODirConfig,
+	isqlDB isql.DB,
+	username username.SQLUsername,
+	externalConnTestingKnobs cloud.ExternalConnTestingKnobs,
+	serverCfg *execinfra.ServerConfig,
+) ExternalConnEnv {
+	return ExternalConnEnv{
+		Settings:                       settings,
+		Conf:                           conf,
+		Isqldb:                         isqlDB,
+		ExternalConnectionTestingKnobs: externalConnTestingKnobs,
+		Username:                       username,
+		ServerCfg:                      serverCfg,
+	}
+}
+
+var _ cloud.KMSEnv = &ExternalConnEnv{}
+
+func (s *ExternalConnEnv) ClusterSettings() *cluster.Settings {
+	return s.Settings
+}
+
+func (s *ExternalConnEnv) KMSConfig() *base.ExternalIODirConfig {
+	return s.Conf
+}
+
+func (s *ExternalConnEnv) DBHandle() isql.DB {
+	return s.Isqldb
+}
+
+func (s *ExternalConnEnv) User() username.SQLUsername {
+	return s.Username
+}
+
+func (s *ExternalConnEnv) ExternalStorageFromURI() cloud.ExternalStorageFromURIFactory {
+	return s.ServerCfg.ExternalStorageFromURI
+}
+
+func (s *ExternalConnEnv) ExternalConnTestingKnobs() cloud.ExternalConnTestingKnobs {
+	return s.ExternalConnectionTestingKnobs
+}
+
+// ValidationFnWithCloud is the type taken by Register*ValidationWithCloud. Validation functions can ping the external connection.
+type ValidationFnWithCloud func(ctx context.Context, env cloud.KMSEnv, uri string) error
```

Comment:
Sorry for repeting myself thrice but just for clarity I think all validation functions should take in an ExternalConnEnv. It is then up to them to construct internal state required for the actual validation.